test: Update E2E utilities

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Related PRs
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6281

## What?
<!-- In a few words, what is the PR actually doing? -->
Update E2E utilities for Appium 2 and WebdriverIO.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Improve the stability and quality of the automated tests.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Set static device pixel ratio
 WebdriverIO (the replacement for the deprecated WD driver library) does
 not have a method for retrieving a device's pixel ratio. Therefore, this
 sets an explicit value for each of our test devices.
- Update the `getAddBlockButton` utility


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
See https://github.com/wordpress-mobile/gutenberg-mobile/pull/6281. 

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
n/a

## Screenshots or screencast <!-- if applicable -->
n/a

dcalhoun committed:
test: Set static device pixel ratio

WebdriverIO (the replacement for the deprecated WD driver library) does
not have a method for retrieving a device's pixel ratio. Therefore, this
sets an explicit value for each of our test devices.

dcalhoun committed:
test: Update the `getAddBlockButton` utility

Update syntax for WebdriverIO, which is a part of the Appium 2 upgrade.

dcalhoun committed:
test: Update toggleDarkMode utility

Seems this method has changed or been removed for Android.

dcalhoun committed:
test: Increase robustness of waitForKeyboardToBeHidden utility

This utility appeared to fail occasionally due to an empty `addButton`
value.

dcalhoun committed:
test: toggleDarkMode uses consistent approach across platforms

This updates the iOS variant to mirror the structure of the Android
variant in the name of consistency.

dcalhoun commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -302,7 +301,10 @@ class EditorPage {
 	async waitForKeyboardToBeHidden() {
 		const { addButtonLocation } = this.initialValues;
 		const addButton = await this.getAddBlockButton();
-		const location = await addButton.getLocation();
```
Occasionally, this query would fail with Appium 2 with the following error. Guarding against an empty query return avoids this error and results in the recursive loop of `waitForKeyboardToBeHidden` to occur. 

```
TypeError: Cannot read properties of undefined (reading 'getLocation')
```

geriux commented on packages/react-native-editor/__device-tests__/helpers/caps.js:
```diff
@@ -11,6 +11,7 @@ const ios = {
 exports.iosLocal = ( { iPadDevice = false } ) => ( {
 	...ios,
 	deviceName: ! iPadDevice ? 'iPhone 13' : 'iPad Pro (9.7-inch)',
+	devicePixelRatio: ! iPadDevice ? 3 : 2,
```
This approach looks good to me while there's no support for the iOS side. If we ever change the simulators, we can tweak this value manually.

For consideration maybe we can use the `pixelRatio` name that the [Android driver uses](https://github.com/appium/appium-uiautomator2-driver/blob/35e55d5ccd72a871c830cacc8b62f00bdd9d4a76/lib/driver.js#L212).

dcalhoun replied:
Addressed in https://github.com/WordPress/gutenberg/pull/55322/commits/5d684e20780e3666aca0b66b763802160a5e6582.

geriux reviewed (approved):
LGTM! 🚀 Nice work!

dcalhoun committed:
refactor: Rename devicePixelRatio to pixelRatio

Mirror Android implementation found in Appium 2.

dcalhoun commented:
Debugging why the tests began failing. Both the [PHP](https://github.com/WordPress/gutenberg/actions/runs/6534670480/attempts/3) and [Android](https://github.com/WordPress/gutenberg/actions/runs/6534670479/attempts/3) failures are odd given the only change since the [last successful build](https://github.com/WordPress/gutenberg/actions/runs/6500760773/attempts/1) was iOS-only testing code: https://github.com/WordPress/gutenberg/pull/55322/commits/5d684e20780e3666aca0b66b763802160a5e6582. I am unable to reproduce the Android failures locally so far.

geriux commented:
> Debugging why the tests began failing. Both the [PHP](https://github.com/WordPress/gutenberg/actions/runs/6534670480/attempts/3) and [Android](https://github.com/WordPress/gutenberg/actions/runs/6534670479/attempts/3) failures are odd given the only change since the [last successful build](https://github.com/WordPress/gutenberg/actions/runs/6500760773/attempts/1) was iOS-only testing code: [5d684e2](https://github.com/WordPress/gutenberg/commit/5d684e20780e3666aca0b66b763802160a5e6582). I am unable to reproduce the Android failures locally so far.

Oh that's weird! Maybe enabling more logging could help to debug, [commenting this](https://github.com/WordPress/gutenberg/blob/5d684e20780e3666aca0b66b763802160a5e6582/packages/react-native-editor/__device-tests__/helpers/utils.js#L177) or changing the level of logging could help.

dcalhoun committed:
test: Increase logging levels

Attempt to debug CI failures.

```
FAIL __device-tests__/gutenberg-editor-rendering-media-blocks.test.js (73.496 s)
  ● Gutenberg Editor Rendering Media Blocks test › should be able to render blocks correctly

    unknown method: Not implemented

      at getErrorFromResponseBody (file:/Users/runner/work/gutenberg/gutenberg/node_modules/webdriver/build/utils.js:194:12)
      at NodeJSRequest._request (file:/Users/runner/work/gutenberg/gutenberg/node_modules/webdriver/build/request/index.js:164:23)
      at Browser.wrapCommandFn (file:/Users/runner/work/gutenberg/gutenberg/node_modules/@wdio/utils/build/shim.js:81:29)
```

dcalhoun committed:
refactor: Prefer `action` to `touchAction`

Avoid the deprecated `touchAction` command.

> The `touchAction` command is deprecated and will be removed in a future version. We recommend to use the action command instead with pointer type touch, e.g.:
>
> await browser.action('pointer', {
>   parameters: { pointerType: 'touch' }
> })

https://webdriver.io/docs/api/element/touchAction/

This may also relate to the fact that Mobile JSON Wire Protocol is
deprecated as well.

> Mobile JSON Wire Protocol
>
> The Mobile JSON Wire Protocol is a super set of mobile commands on top of the JSON Wire Protocol. Given this one is deprecated the Mobile JSON Wire Protocol also got deprecated. Appium might still support some of its commands but it is not recommended to use them.

https://webdriver.io/docs/api/protocols#mobile-json-wire-protocol

dcalhoun committed:
Revert "test: Increase logging levels"

This reverts commit 7bb46decef07d4ff3c4956d717e291a2d8222b8e.

dcalhoun commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -450,26 +450,16 @@ const swipeFromTo = async (
-			options: {},
-		},
-	] );
+	await driver
+		.action( 'pointer', {
+			parameters: { pointerType: 'touch' },
+		} )
+		.move( { ...from, duration: 0 } )
+		.down( { button: 0 } )
+		.move( { ...to, duration: 300 } )
+		.up( { button: 0 } )
+		.pause( delay )
+		.perform();
```
As outlined in https://github.com/WordPress/gutenberg/pull/55322/commits/bf527eec76635a8a6eaf1fd2e441ed836487acdc, this refactors to avoid the deprecated `touchPerform` command. It was done to address recent [CI task failures](https://github.com/WordPress/gutenberg/pull/55322#issuecomment-1764742815). It is unclear why the usage of this command caused seemingly sporadic test failures, but it is likely good to avoid this regardless. 

@geriux thank you for the [note on log levels](https://github.com/WordPress/gutenberg/pull/55322#issuecomment-1764767359). It allowed me to observe the failure occurring immediately after this `touchPerform` call. 🙇🏻

geriux replied:
Glad it helped! And TIL about the deprecated command! Thanks for making the changes 🚀

dcalhoun committed:
test: Set default `swipeFromTo` delay value

This prevents `pause` from being called without a parameter, which
throws an error.

dcalhoun commented:
Still do not know why the PHP tests fail, but given this does not target the main branch, I am going to merge this to unblock other work. 

@geriux please feel free to share any additional change requests. I'd be happy to follow up as needed. Thanks again for the help.
